server/src/main/java/io/crate/execution/ddl/tables/MappingUtil.java

```diff
@@ -204,4 +219,52 @@ public static void mergeConstraints(Map<String, Object> meta,
             }
         }
     }
+
+    /**
+     * @param columnsMap is "properties" map of the index metadata.
+     *
+     * For DDL-s adding leaves to existing objects columns, non-leaf columns have OID assigned.
+     * For DML-s it could be that OID is already assigned in case of concurrent dynamic mapping updates.
+     * Hence, we must skip such columns and also increment only when needed.
+     *
+     * @return new columnOID value which MUST BE updated in the cluster state by the caller.
+     */
+    public static long assignOID(@Nullable Map<String, Map<String, Object>> columnsMap, long currentOID) {
```

Comment:
What do you think of integrating the oid assignment to `createPropertiesMap` / `addColumnProperties`, or even add a oid parameter to `Reference#toMapping`?

I think it would avoid spreading know-how of the mapping structure and it's also more efficient as we'd avoid additional hashmap lookups. It also seems a bit odd to first create a structure with missing data, only to fill it in immediately afterwards.

Adding it to `toMapping` would help enforce the use of a oid if a mapping is created.

Reply:
> avoid spreading know-how of the mapping structure and it's also more efficient as we'd avoid additional hashmap lookups

good point. 

oid is already in `SimpleReference.toMapping` but to be fair, it does nothing, it's `assignOID` doing that.

I think I should set oid in Reference (not sure which way is better)
1. oid is final and new method `assignOid` creates a clone, I did smth like that in the old draft in https://github.com/crate/crate/pull/13757/commits/d0dc7e0988752d8b5043fb38c7db88952e126de5
2. oid is non-final and just use setter (not that nice but can avoid cloning - maybe less burden for dynamic mapping updates. We already use "some fields final and some not" in `GeneratedReference`. wdyt?

Then I will get rid of `assignOID` and after normalizing columns will just assign OID in a simple loop like
```
request  = normalizeColumn(request) // adds parents and gets latest refs from the cluster state
maxOID = clusteState.metdata.columnOID();
for (Ref ref: request.referneces) {
 maxOID = ref.assignOID(maxOID);
}
```
and then existing already in `toMapping`  call of `mapping.put("oid", oid)` will actually do smth

Reply:
> Just pass the actual oid along to toMapping or integrate the logic to addColumnProperties

will do one of those 

on swapping `DynamicReference/Reference` - replied in https://github.com/crate/crate/pull/14142#issuecomment-1550981418